insights_connexion/app.py

```diff
@@ -32,6 +39,65 @@ def validate_schema(self, data, url):
         self.validator.validate(data)
 
 
+@web.middleware
+async def error_middleware(request, handler):
+    try:
+        response = await handler(request)
+    except(ValidationError) as e:
+        log.error(e)
+        response = responses.invalid_request_parameters()
+    except(UniqueViolationError) as e:
+        log.error(e)
+        response = responses.resource_exists()
+    except(NoResultFound) as e:
+        log.error(e)
+        response = responses.not_found()
+    except(Exception) as e:
+        log.error(e)
+        response = responses.internal_server_error()
+
+    return response
+
+
+def _items_to_json(items):
+    return {k: v for k, v in items if v is not None}
+
+
+@web.middleware
+async def log_middleware(request, handler):
+    try:
+        req_body = await request.json() if request.has_body else None
+    except(JSONDecodeError) as e:
+        log.error(e)
+        return responses.invalid_request_parameters('Malformed JSON in request body.')
+
+    req_id = str(uuid.uuid4())
+
+    log.info(json.dumps({'type': 'request',
+                         'req_id': req_id,
+                         'body': req_body,
+                         'cookies': _items_to_json(request.cookies.items()),
+                         'content-type': request.content_type,
+                         'content-length': request.content_length,
+                         'headers': _items_to_json(request.headers.items()),
+                         'method': request.method,
+                         'query': _items_to_json(request.query.items()),
+                         'url': str(request.url)}))
+
+    response = await handler(request)
+
+    log.info(json.dumps({'type': 'response',
+                         'req_id': req_id,
+                         'body': response.text,
```

Comment:
Hmm. I see that the Response doesn’t pertain the original JSON object. It can be parsed though. If we send only JSON responses and we know they are valid, e.g. because they’re populated by _dumps_ from Python objects, then it can be as simple as

```python
body = loads(response.text)
```

Otherwise there would have to be some more logic. Not all parts are necessary depending on what responses we actually emit.

```python
body = response.text or None
if response.content_type == 'application/json':
    try:
        body = loads(response.text)
    except ValueError:
        pass
```

The only caveat I see in the simplest solution is that if we start to emit non-JSON or empty responses, then logging would start raising an error. But that’s what tests are for.